### ISSUE TITLE
Creation of buckets, accounts, and policies as custom resources

### ISSUE THREAD

**Issue by vbr-cts** [minio/operator#1100]
**Is your feature request related to a problem? Please describe.**

I use minio as a backend for a highly multi-tenant SaaS app. We rely on fine-grained policies to make sure each service only accesses whatever it is needs to access. Currently, we provision, manage and revoke these accesses by calling the created tenant's API. This has several drawbacks:

 - we maintain provisioning scripts outside our classic helm workflow
 - disaster recovery is complicated as we need to re-create all ACLs and accounts
 - developer experience and CI is complicated as we can't spin up an entire clean env from a helm chart
 - modifying existing policies is complex and error-prone

**Describe the solution you'd like**

I'd like to have something where I can have custom resources for:

 - buckets
 - policies
 - service accounts

These would be linked to a minio tenant by name.

Ideally, the service account CRDs would point to a secret by namespace and name, and the key would be stored there (or read from there if it already exists).

This way, when deploying a new tenant, I can have helm charts generate a MinioBucket, a MinioServiceAccount and the required MinioPolicies onto a pre-existing minio tenant. The services would wait for the secret described in the MinioServiceAccount to be populated by the operator before starting.

Similarly:

 - creating/deleting the MinioBucket does just that
 - modifying a MinioPolicy updates the policies of linked buckets/service accounts

**Describe alternatives you've considered**

 - bunch of scripts: currently doing this, not great
 - s3-operator (https://github.com/agill17/s3-operator) might be an option, but it is unmaintained, and having it play nicely with minio-operator's multi-tenancy is proving really hard, it seems like we need to deploy a separate operator for each minio tenant.
 - integration within our own (in-house) operator: can work, but we'd rather have

**Comment by MarkusBauerBE** [comment]
Seconding this - iirc the reason this wasn't implemented was because it was to be outsourced. 
I've personally solved this by running a K8s job that scans the content of a configmap. This is then deployed via Helm and a Helm hook triggers the job to run, whenever the CM gets updated (only adding the buckets that don't already exist).
It works, but this should have frankly been implemented as a crd a long long time ago.

**Comment by mbrancato** [comment]
I mentioned this in the Slack channel prior to seeing this, but I think this would definitely further align this operator with Kubernetes principles of declarative configuration. As it currently stands, I think the Operator console will sync its tenant changes back to the cluster as `Tenant` resources. This would be extremely useful to have a similar setup when a bucket, user, policy, etc is created.

**Comment by mbrancato** [comment]
I wanted to add some us-case context here based on what I've read is required for the operator today. For me, I almost never do a `kubectl apply`, let along deploying a `kubectl` plugin for Minio and then running `kubectl minio init` or something to install. There is also a helm chart, which I can work with easily on a declarative basis using the [HelmRelease](https://fluxcd.io/docs/components/helm/api/) resource. Even in a testing environment, which I'm trying to deploy here, I don't do anything manually. We deploy all kubernetes configurations and resources using GitOps, and nobody is directly applying resources. 

To get a little deeper in this, when I consider using a tool, I start to think about reproducibility. Can I tear this environment down and stand it back up easily? Outside a few small things related to bootstrapping an environment, there are no manual commands run. Infrastructure is built using automated tools and then an in-cluster CD operator like flux picks up the kubernetes resources configs, and the principle of eventual consistency takes hold. Take for example, I want to deploy a new environment, configure it, and that environment has some app that uses a Minio bucket. Where this starts to break down is we would have failures in rebuilding the environment without basically a second bootstrapping step to run some scripts to create the users, policies, and buckets.

**Comment by addisonautomates** [comment]
Please support using configmaps and secrets for scrubbing to provision buckets and policies after initial deployment.  This is the only tool we run on kubernetes that I can't just use Flux to instantly get to a fully configured state.  Having to fire kubernetes jobs to run scripts is an anti-pattern that kubernetes and progressive deployment tools are trying to avoid.  Take a look at how grafana handles the creation of new folders and dashboards to get a good idea how to organize things.  Would love to be able to add to a master configmap for bucket creation or be able to add new configmaps with bucket references and have minio pick them up as long as a certain kubernetes label is used

**Comment by MyIgel** [comment]
That is still an issue :)

**Comment by WoodyWoodsta** [comment]
I'm here for this as well. I'm not entirely sure what the use is of the Tenant CRD as a k8s IaC resource object if there isn't the ability to manage the tenant via this object after provisioning.

**Comment by sathieu** [comment]
You can manage buckets and policies with: https://github.com/InseeFrLab/s3-operator

**Comment by kamikaze** [comment]
+1

**Comment by allanrogerr** [comment]
This is currently being worked onWe are working on a declarative approach to minio configuration through custom resources. We will update when there is something to announce.

**Comment by danieljkemp** [comment]
This is one of the features I'm missing from other operators (like cnpg). In particular the user credential management. If no credentials are specified, the operator should populate the secret with credentials.

**Comment by harshavardhana** [comment]
@allanrogerr @pjuarezd this issue can be closed ?

**Comment by allanrogerr** [comment]
For this feature and others, please review our new releases periodically. Whenever we make a release, you can see that on https://github.com/minio/minio/releases
You may also subscribe to our RSS feed: https://github.com/minio/minio/releases.atom

How to: https://medium.com/@vilcins/rss-feeds-for-your-github-releases-tags-and-activity-cbda2c51373

**Comment by dhess** [comment]
@harshavardhana @allanrogerr I assume this issue was closed because it's now possible to create buckets, accounts, and policies as custom resources. That's great news!

However, after reviewing this project's new releases, as suggested by the parent comment that closed the issue, I'm afraid I can't find any reference to these new features. Perhaps @harshavardhana or @allanrogerr could help me out and point out the new CRDs that I'm missing?

**Comment by danieljkemp** [comment]
Will operator CRDs be in the release notes of Minio, or the Minio Operator releases? The later has release notes as a simple unsorted list of commits which is less helpful. The former is a stream of bugfixes.

If there is another issue tracking development of the feature, or a WIP PR, that would be helpful for tracking rather than expecting watching a stream of release notes that are mostly bug fixes and then searching the non-bug fixes for a feature without knowing when it might even be released.

**Comment by benedikt-bartscher** [comment]
Any hints which release contains those features or where to find more information about the new CRDs?

**Comment by Skaronator** [comment]
It's been a month since this issue was closed, and I still don't find any relevant Changelogs or Code changes for this feature. Therefore, this issue should remain open.

I would expect changes in the CRD, but nothing so far popped up: https://github.com/minio/operator/commits/f8831e30f32050bf8d48e8c299b5d94c116a7525/resources/base/crds

**Comment by tentakle** [comment]
Why was this issue closed if nothing was done?

**Comment by pjuarezd** [comment]
I will share how to it soon

**Comment by benedikt-bartscher** [comment]
@mikkeschiren thanks, this seems to work fine.
However, a separate CRD for buckets which holds a ref to the minio tenant cr would be much more useful.

**Comment by mikkeschiren** [comment]
Agree on that @benedikt-bartscher - I like the way Grafana is handling configuration for dashboards, datasources etc.

**Comment by fgarciacode** [comment]
Do you have an example how tenant's service accounts can be created?

**Comment by benfiola** [comment]
Out of both necessity and curiosity, I ended up writing [my own operator](https://github.com/benfiola/minio-operator-ext) to manage MinIO resources. 

It's pretty simple (as my use case is pretty simple) -but maybe someone else will find this useful.

In any case - keeping my eyes peeled on this ticket!

**Comment by john-funk** [comment]
> I will share how to it soon

any updates?

**Comment by ramondeklein** [comment]
There is not a lot of documentation yet, but you can already use [job.min.io/v1alpha1](https://github.com/minio/operator/blob/master/docs/job_crd.adoc#jobminiov1alpha1) to submit jobs to Minio via Kubernetes CRDs with the latest version of the operator. Note that this functionality is still in alpha and may change.

This will create a service account (named `mc-job-sa`) that is binds the SA to the `consoleAdmin` policy and creates a bucket in the `myminio` tenant:
```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: mc-job-sa
  namespace: minio-operator
---
apiVersion: sts.min.io/v1alpha1
kind: PolicyBinding
metadata:
  name: mc-job-binding
  namespace: minio-operator
spec:
  application:
    serviceaccount: mc-job-sa
    namespace: minio-operator
  policies:
    - consoleAdmin
---
apiVersion: job.min.io/v1alpha1
kind: MinIOJob
metadata:
  name: minio-test-job
  namespace: minio-operator
spec:
  serviceAccountName: mc-job-sa
  tenant:
    name: myminio
    namespace: minio-operator
  commands:
    - op: make-bucket
      args:
        name: memes
```

**Comment by danieljkemp** [comment]
My goal with such a feature is to be able to create buckets/policies/users via CRDs and have the operator manage them.

I'd hope to be able to create a bucket and user via CRDs, with the operator populating a token for the user that can be consumed by an application that uses the bucket, without me needing to manage that secret myself

**Comment by iegurbide** [comment]
Thank you for the example @ramondeklein.

Is there a description of available op and required args for the command section?

Your example make-bucket doen't seem to match any of the documented APIs.

**Comment by ramondeklein** [comment]
@iegurbide We're in the process of documenting this functionality, but it's not completely ready yet. I also wouldn't recommend this at this stage for production.

Some pointers:
- https://github.com/minio/operator/pull/2173
- https://github.com/minio/operator/blob/master/docs/job_crd.adoc

**Comment by mx-sylveon** [comment]
My two cents - you guys are trying to reinvent the wheel here and not at all building things the k8s way.

This comment from 2022 is pretty accurate regarding the anti-patterns going on: https://github.com/minio/operator/issues/1100#issuecomment-1286059866

Why, with your currently in-progress solution, are we forced to configure MinIO via a generic "MinIOJob" CRD?

Look at pretty much any other operator project and you should know that something feels off here.

Just wrapping MinIO configuration in a generic Job CRD instead of actually building resources like Bucket, Policy, etc; is frankly a pretty lazy way to go about things and is very unclear to users when compared to the alternatives.

The point of operators is for them to manage the lifecycle of an application's configuration and resources based on a declarative definition - not provide a wrapper for running one-off CLI commands.

With the "make-bucket" example given, what if I want to delete that bucket? Deleting the MinIOJob presumably does nothing, as that just says "run this command once(?)". At least I assume it just runs the command once.

The more I look at this, in particular the [Job CRD linked](https://github.com/minio/operator/blob/master/docs/job_crd.adoc) and how the proposed solution involves specifying MinIO Client commands, complete with the ability to define for them arrays of other commands they depend on via a `dependsOn` property... the more I feel like you're basically trying to build your own IaC tooling akin to Terraform instead of using the tools that have been built for you.

I strongly encourage taking a step back and not getting stuck in the sunk-cost fallacy on this given the work already completed.

**Comment by rmvangun** [comment]
@benfiola Your CRD spec for these resources seems really obvious, and is, I believe, what most people expect when installing the minio-operator. In fact even ChatGPT, when describing MinIO implementations, actually hallucinates these kinds of CRDs, it's pretty funny.

Would you be willing to contribute that back to this project?

**Comment by benfiola** [comment]
@rmvangun ultimately this decision is up to the maintainers of minio + minio-operator - but i'm more than happy to contribute part of or all of my work to the upstream project.  i'm also more than happy to implement the reconciliation logic for these schemas as well.

@ramondeklein - would you be receptive to me contributing this feature to the project?  (for reference, here's the [crd schema](https://github.com/benfiola/minio-operator-ext/blob/dev/manifests/crds.yaml) and [example usage](https://github.com/benfiola/minio-operator-ext/blob/dev/manifests/example-resources.yaml)

**Comment by ramondeklein** [comment]
We don't plan to implement a CRD-based management of buckets, users, policies, ... into the operator in the near future. We currently don't have a strong request for this feature and it will take up too much of our resources. It's not only the initial coding, but it also needs documentation, automated tests and most importantly maintenance and support. The MinioJob functionality may not meet everyone's expectations, but it allows us to perform all kind of tasks via mc in a somehow automated way with limited effort. Of course, the community can create a tenant-level operator to implement the required functionality. But for the reasons listed above, we won't integrate it into the operator.

**Comment by pierreozoux** [comment]
As an infrastructure operator, my wish is to declare the infra I need for my apps.

Usually, for an app, I need a user, a bucket and a policy associated.

I don't want to Create or Update a bucket, I want the bucket to be there, and configured the way I said.

In kubernetes world, we expect declarative APIs, and in minio case, we would expect a declarative bucket resource (CRD). The nice thing is that you can put this resource in a git repo as well.

`mc` is an imperative API.

So, yes `mc`, or minio API is there for all imperative tasks, and it works great.

The idea of the feature request is to get a native kubernetes declarative way to manage minio resources.

I understand that you are not interested to develop it, I just fear you didn't really understand the feature request. The last comment made me react, I'm not sure you really grasp the difference between declarative and imperative API.

**Comment by benfiola** [comment]
🫡 - sounds good and thank you for the response @ramondeklein.  feel free to ping me if you change your mind - would love to implement this!  

meanwhile, if anyone else is interested, feel free to give my [operator](https://github.com/benfiola/minio-operator-ext) a shot.  it's not perfect (i wrote it, heh) - but maybe it'll work for you.  if not, let me know and maybe i can fix it.

**Comment by bauerjs1** [comment]
I understand that developer resources are limited and that this feature request would probably require some and ofc I respect the current decision. I just want to give another two cents and +1 for the feature request.

From my personal perspective, I am currently evaluating different storage solutions for Kubernetes. Having minio advertised as "Kubernetes-native object storage" I actually assumed it to have a full-featured operator allowing me to install, manage and configure the storage declaratively via CRDs. I am very interested in using Minio but not having this feature is probably a no-go for my use case. I am also a bit surprised that there is little demand for this – how does management effort scale without a declarative API? Can users/developers request their buckets and receive the credentials in a self-service manner (may be an off-topic question here)?

I would also strongly prefer CRDs (control loops) over Terraform (again, one-shot job runs), if Minio is considered to be K8s-native.

Maybe I'll give Minio a try with your additional operator @benfiola. Thanks a lot for taking the effort!

**Comment by batazor** [comment]
If I use `kind: MinIOJob` how set access_key/access_secret for a bucket?

**Comment by ramondeklein** [comment]
@batazor A bucket doesn't have an access/secret key, but these are used to access MinIO instances. It's also not related to this issue.

**Comment by jochenseeber** [comment]
> You can manage buckets and policies with: https://github.com/InseeFrLab/s3-operator

Still would be nice if we didn't have to fall back to another solution...
